File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/ab064721cbb4c4373b752374fc4f664d.py
import sys
import random
from math import acos

INF = int(1e9)
INF64 = int(1e18)
MOD = INF + 7
EPS = 1e-9
PI = acos(-1.0)

N = 2000 + 7

class Point:
    def __init__(self, x=0, y=0, i=-1):
        self.x = x
        self.y = y
        self.i = i

    def __lt__(self, other):
        if self.x != other.x:
            return self.x < other.x
        return self.y < other.y

    def __sub__(self, other):
        return Point(self.x - other.x, self.y - other.y)

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

def sqr(a):
    return a * a

def sdist(a, b):
    return sqr(a.x - b.x) + sqr(a.y - b.y)

def dot(a, b):
    return a.x * b.x + a.y * b.y

def cross(a, b):
    return a.x * b.y - a.y * b.x

def build(points, n):
    if n <= 1:
        return points[:n]

    up = [points[0]]
    dw = [points[0]]

    for i in range(1, n):
        if i == n - 1 or cross(points[n - 1] - points[0], points[i] - points[0]) >= 0:
            while len(up) > 1 and cross(points[i] - up[-2], up[-1] - up[-2]) < 0:
                up.pop()
            up.append(points[i])
        if i == n - 1 or cross(points[n - 1] - points[0], points[i] - points[0]) < 0:
            while len(dw) > 1 and cross(points[i] - dw[-2], dw[-1] - dw[-2]) > 0:
                dw.pop()
            dw.append(points[i])

    res = up + dw[-2:0:-1]
    return res

def read():
    n = int(sys.stdin.readline().strip())
    points = []
    for i in range(n):
        x, y = map(int, sys.stdin.readline().strip().split())
        points.append(Point(x, y, i + 1))
    s = sys.stdin.readline().strip()
    return points, n, s

def solve(points, n, s):
    points.sort()
    ans = []

    ch = build(points, n)
    ans.append(ch[0])
    if s[0] == 'L':
        ans.append(ch[-1])
    else:
        ans.append(ch[1])

    for i in range(len(s)):
        x = points.index(ans[-2])
        points = points[:x] + points[x + 1:]
        n -= 1
        ch = build(points, n)
        pos = ch.index(ans[-1])
        if s[i + 1] == 'L':
            ans.append(ch[(pos - 1 + len(ch)) % len(ch)])
        else:
            ans.append(ch[(pos + 1) % len(ch)])

    print(' '.join(str(it.i) for it in ans))

def main():
    points, n, s = read()
    solve(points, n, s)

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/ab064721cbb4c4373b752374fc4f664d.go
package main

import (
	"fmt"
	"math"
	"sort"
)

type Point struct {
	x, y, i int
}

var n int
var a []Point
var s string

func read() bool {
	if _, err := fmt.Scan(&n); err != nil {
		return false
	}
	a = make([]Point, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i].x, &a[i].y)
		a[i].i = i + 1
	}
	fmt.Scan(&s)
	return true
}

func (p Point) subtract(q Point) Point {
	return Point{p.x - q.x, p.y - q.y, -1}
}

func (p Point) add(q Point) Point {
	return Point{p.x + q.x, p.y + q.y, -1}
}

func (p Point) equals(q Point) bool {
	return p.x == q.x && p.y == q.y
}

func sqr(a int64) int64 {
	return a * a
}

func sdist(a, b Point) int64 {
	return sqr(int64(a.x-b.x)) + sqr(int64(a.y-b.y))
}

func dot(a, b Point) int64 {
	return int64(a.x)*int64(b.x) + int64(a.y)*int64(b.y)
}

func cross(a, b Point) int64 {
	return int64(a.x)*int64(b.y) - int64(a.y)*int64(b.x)
}

func build() []Point {
	if n <= 1 {
		return a[:n]
	}

	up := []Point{a[0]}
	dw := []Point{a[0]}

	for i := 1; i < n; i++ {
		if i == n-1 || cross(a[n-1].subtract(a[0]), a[i].subtract(a[0])) >= 0 {
			for len(up) > 1 && cross(a[i].subtract(up[len(up)-2]), up[len(up)-1].subtract(up[len(up)-2])) < 0 {
				up = up[:len(up)-1]
			}
			up = append(up, a[i])
		}
		if i == n-1 || cross(a[n-1].subtract(a[0]), a[i].subtract(a[0])) < 0 {
			for len(dw) > 1 && cross(a[i].subtract(dw[len(dw)-2]), dw[len(dw)-1].subtract(dw[len(dw)-2])) > 0 {
				dw = dw[:len(dw)-1]
			}
			dw = append(dw, a[i])
		}
	}

	res := up
	for i := len(dw) - 2; i >= 1; i-- {
		res = append(res, dw[i])
	}

	return res
}

func solve() {
	sort.Slice(a, func(i, j int) bool {
		if a[i].x != a[j].x {
			return a[i].x < a[j].x
		}
		return a[i].y < a[j].y
	})

	ans := []Point{}
	ch := build()
	ans = append(ans, ch[0])
	if s[0] == 'L' {
		ans = append(ans, ch[len(ch)-1])
	} else {
		ans = append(ans, ch[1])
	}

	for i := 0; i < len(s); i++ {
		x := sort.Search(len(a), func(j int) bool {
			return a[j].equals(ans[len(ans)-2])
		})
		copy(a[x:], a[x+1:])
		n--

		ch = build()
		pos := sort.Search(len(ch), func(j int) bool {
			return ch[j].equals(ans[len(ans)-1])
		})

		if s[i+1] == 'L' {
			ans = append(ans, ch[(pos-1+len(ch))%len(ch)])
		} else {
			ans = append(ans, ch[(pos+1)%len(ch)])
		}
	}

	for _, it := range ans {
		fmt.Printf("%d ", it.i)
	}
	fmt.Println()
}

func main() {
	if read() {
		solve()
	}
}
